pkg/webhook/admission/pod/metrics_aggregate_injector.go

```diff
@@ -17,46 +17,90 @@ limitations under the License.
-				pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{Name: constants.QueueProxyAggregatePrometheusMetricsPortEnvVarKey, Value: constants.QueueProxyAggregatePrometheusMetricsPort})
-
-				// If SetPrometheusAggregateAnnotation is true, the pod annotations for prometheus port and path will be set. The scrape annotation is not set,
-				// that is left for the user to configure.
-				if setPromAnnotation, ok := pod.ObjectMeta.Annotations[constants.SetPrometheusAggregateAnnotation]; ok && setPromAnnotation == "true" {
-					pod.ObjectMeta.Annotations[constants.PrometheusPortAnnotationKey] = constants.QueueProxyAggregatePrometheusMetricsPort
-					pod.ObjectMeta.Annotations[constants.PrometheusPathAnnotationKey] = constants.DefaultPrometheusPath
-				}
+			// The kserve-container prometheus port/path is inherited from the ClusterServingRuntime YAML.
+			// If no port is defined (transformer using python SDK), use the default port/path for the kserve-container.
+			kserveContainerPromPort := defaultKserveContainerPrometheusPort
+			if port, ok := pod.ObjectMeta.Annotations[constants.KserveContainerPrometheusPortKey]; ok {
+				kserveContainerPromPort = port
+			}
+
+			kserveContainerPromPath := constants.DefaultPrometheusPath
+			if path, ok := pod.ObjectMeta.Annotations[constants.KServeContainerPrometheusPathKey]; ok {
+				kserveContainerPromPath = path
 			}
+
+			// The kserve container port/path is set as an EnvVar in the queue-proxy container
+			// so that it knows which port/path to scrape from the kserve-container.
+			pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{Name: constants.KServeContainerPrometheusMetricsPortEnvVarKey, Value: kserveContainerPromPort})
+			pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{Name: constants.KServeContainerPrometheusMetricsPathEnvVarKey, Value: kserveContainerPromPath})
+
+			// Set the port that queue-proxy will use to expose the aggregate metrics.
+			pod.Spec.Containers[i].Env = append(pod.Spec.Containers[i].Env, v1.EnvVar{Name: constants.QueueProxyAggregatePrometheusMetricsPortEnvVarKey, Value: constants.QueueProxyAggregatePrometheusMetricsPort})
+
 		}
 	}
+}
+
+// InjectMetricsAggregator looks for the annotations to enable aggregate kserve-container and queue-proxy metrics and
+// if specified, sets port-related EnvVars in queue-proxy and the aggregate prometheus annotation.
+func (ma *MetricsAggregator) InjectMetricsAggregator(pod *v1.Pod) error {
+	//Only set metric configs if the required annotations are set
+	enableMetricAggregation, ok := pod.ObjectMeta.Annotations[constants.EnableMetricAggregation]
+	if !ok {
+		if pod.ObjectMeta.Annotations == nil {
+			pod.ObjectMeta.Annotations = make(map[string]string)
+		}
+		pod.ObjectMeta.Annotations[constants.EnableMetricAggregation] = ma.EnableMetricAggregation
+	} else if enableMetricAggregation == "true" {
+		setMetricAggregationEnvVars(pod)
+	}
```

Comment:
I think we also need to set the env vars when the annotation does not present but is enabled in the configmap
```go
if pod.ObjectMeta.Annotations[constants.EnableMetricAggregation] == "true" {
 	setMetricAggregationEnvVars(pod)
}

```

Reply:
this code does that! see the lines right above

Reply:
ah wait, I need to remove the "else" and just have "if"

Reply:
good point

Reply:
updated

Reply:
I think you would need to use `pod.ObjectMeta.Annotations[constants.EnableMetricAggregation]` in the if condition as `enableMetricAggregation` is only set when the annotation exists, when annotation is absent you add the annotation with the value from configmap.

Reply:
fixed